File: app/src/main/java/com/chichiangho/base/http/RxHttpObserver.java
package com.chichiangho.base.http;

import android.content.Intent;

import com.chichiangho.base.activity.LoginActivity;
import com.chichiangho.base.base.BaseApplication;
import com.chichiangho.base.base.BaseResponse;
import com.chichiangho.base.utils.ToastUtil;

import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;

/**
 * Created by chichiangho on 2017/4/27.
 */

public abstract class RxHttpObserver<T> implements Observer<T> {

    @Override
    final public void onSubscribe(Disposable d) {
        onStart();
    }

    @Override
    final public void onNext(T value) {
        onSuccess(value);//错误码解析在HandleResponse
    }

    @Override
    final public void onError(Throwable e) {
        if (e != null && e.getMessage() != null) {
            switch (e.getMessage()) {
                case BaseResponse.CODE_TOKEN_TIMEOUT:
                    ToastUtil.showShort("token过期请重新登录");
                    Intent intent = new Intent(BaseApplication.getInstance(), LoginActivity.class);
                    intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);
                    BaseApplication.getInstance().startActivity(intent);
                    break;
                default:
                    ToastUtil.showShort(e.getMessage().equals("") ? "网络错误，请稍后再试..." : e.getMessage());
                    break;
            }
        }

        onFailed();
        onComplete();
    }

    protected void onStart() {
    }

    protected abstract void onSuccess(T value);

    protected void onFailed() {
    }

    @Override
    public void onComplete() {
    }
}

File: app/src/main/java/com/chichiangho/base/http/RxHttpObserver.kt
package com.chichiangho.base.http

import com.chichiangho.base.base.BaseResponse
import com.chichiangho.base.extentions.toastShort
import io.reactivex.Observer
import io.reactivex.disposables.Disposable
import java.net.SocketTimeoutException

/**
 * Created by chichiangho on 2017/4/27.
 */

abstract class RxHttpObserver<T> : Observer<T> {

    override fun onSubscribe(d: Disposable) {}

    override fun onNext(value: T) {
        onSuccess(value)//错误码解析在HandleResponse
    }

    override fun onError(e: Throwable) {
        e.message?.let {
            when (e.message) {
                BaseResponse.CODE_TOKEN_TIMEOUT -> toastShort("token过期请重新登录")
                else -> {
                    if (e.message?.contains("Failed to connect") == true) {
                        toastShort("网络异常")
                    } else if (e.message != "") {
                        toastShort(e.message as String)
                    }
                }
            }
        } ?: let {
            if (e is SocketTimeoutException)
                toastShort("time out")
        }

        onFailed()
        onComplete()
    }

    protected abstract fun onSuccess(response: T)

    protected fun onFailed() {}

    override fun onComplete() {}
}
